### PR TITLE
ADR 033: UnsupportedType

### PR DESCRIPTION
**Pull request by StephenCathcart**
Bolt 6 introduces a new data type representing a value of a type that the server cannot transmit to the client due to the negotiation of an incompatible bolt protocol version.

### CODE REVIEW

**Comment by StephenCathcart** [comment]
[[Go] Introduce Unknown Type to bolt 6.0](https://trello.com/c/URvpgwWp/6401-go-introduce-unknown-type-to-bolt-60)

File: neo4j/dbtype/unsupported.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+// version that supports it.
+type UnsupportedType struct {
+	Name                   string
+	MinimumProtocolVersion db.ProtocolVersion
+	Message                *string
+}
+
+// String returns a string representation of the UnsupportedType.
+func (u *UnsupportedType) String() string {
+	return fmt.Sprintf("UnsupportedType<%s>", u.Name)
```
Comment (robsdedude) [inline neo4j/dbtype/unsupported.go]: Have you considered `"UnsupportedType[%s]"` to be more in line with Go's generic syntax? Not saying it's better, it's just where my mind went first.

File: neo4j/dbtype/unsupportedtypes_test.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/neo4j/neo4j-go-driver/v6/neo4j/db"
+)
+
+func TestUnsupportedTypes(t *testing.T) {
+	t.Run("String representation of UnsupportedType", func(t *testing.T) {
+		unsupported := &UnsupportedType{
+			Name:                   "UUID",
+			MinimumProtocolVersion: db.ProtocolVersion{Major: 6, Minor: 0},
+			Message:                nil,
```
Comment (robsdedude) [inline neo4j/dbtype/unsupportedtypes_test.go:31]: Suggestion: add another test with a non-nil message to make sure the message doesn't show up in the string representation.

File: neo4j/internal/bolt/hydrator_test.go [hunk 3]
```diff
@@ -1132,6 +1187,11 @@ func TestHydrator(outer *testing.T) {
 	}
 }
 
+// Helper function to create string pointer
+func stringPtr(s string) *string {
+	return &s
+}
```
Comment (robsdedude) [inline neo4j/internal/bolt/hydrator_test.go]:
```suggestion
// Helper function to create a pointer (e.g., for an untyped literal)
func ptr[T any](x T) *T {
	return &x
}
```
maybe?

File: testkit-backend/backend.go [hunk 4]
```diff
@@ -1357,6 +1357,7 @@ func (b *backend) handleRequest(req map[string]any) {
 				"Feature:API:Type.Spatial",
 				"Feature:API:Type.Temporal",
 				"Feature:API:Type.Vector",
+				"Feature:API:Type.UnknownType",
```
Comment (robsdedude) [inline testkit-backend/backend.go]: Feature name and TestKit protocol specs will be adjusted to follow the new naming convention (unsupported instead of unknown).

**Review by robsdedude — APPROVED** [review]
:grey_question::writing_hand:

File: neo4j/internal/bolt/hydrator_test.go (hunk 3 above)
Comment (StephenCathcart) [reply neo4j/internal/bolt/hydrator_test.go]: 👍

File: neo4j/dbtype/unsupported.go (hunk 1 above)
Comment (StephenCathcart) [reply neo4j/dbtype/unsupported.go]: Yeah, agree, I just went off ADR, but that makes more sense.

**Comment by StephenCathcart** [comment]
[[Go] Add link to new manuals page to UnsupportedType API Documentation](https://trello.com/c/wZkPZFJN/6416-go-add-link-to-new-manuals-page-to-unsupportedtype-api-documentation)
